[MAG-3278] Redesign Language Dropdown

philjonasfiltered opened the pull request:
# <JIRA issue / Title>

## TL;DR

-   Added a globe icon as per the updated design

## Details
<img width="960" alt="image" src="https://user-images.githubusercontent.com/64406967/101929353-a3107a80-3bce-11eb-9e34-493628ed8bca.png">


Checklist / Reminders:

-   [x] Documentation / Story is created / adapted
-   [x] Unit tests
-   [ ] CSS changes are BEM-compliant

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -69,10 +70,15 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
-            <SvgEdit />
-          </StyledIcon>
-
+          {icon ? (
+            <StyledIcon width={16} height={16} rotate={0}>
+              {icon}
+            </StyledIcon>
+          ) : (
+            <StyledIcon width={16} height={16} rotate={showDropdown ? 270 : 90}>
+              <SvgEdit />
+            </StyledIcon>
+          )}
```
I'd suggest to just pass the actual icon in thats required instead of toggling between them.

```suggestion
            <StyledIcon width={16} height={16} rotate={0}>
              {svgIcon}
            </StyledIcon>
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -26,6 +26,7 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
     showDropdown,
     toggleDropdown,
     curvedWithShadow,
+    icon,
```
```suggestion
    svgIcon,
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -10,6 +10,7 @@ export interface DropdownProps {
   onChange?: (index: number) => void
   options: DropdownOptionProps[]
   curvedWithShadow?: boolean
+  icon?: React.ReactElement
```
```suggestion
  svgIcon?: React.ReactElement
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -99,7 +105,7 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
           </StyledDropdownSelect>
           <StyledIcon width={16} height={16} rotate={90}>
-            <SvgEdit />
+            {icon ? icon : <SvgEdit />}
```
```suggestion
            {svgIcon}
```

matthill8286 commented on src/components/Molecules/LanguageDropDown/LanguageDropdown.tsx:
```diff
@@ -18,6 +19,7 @@ export const LanguageDropdown: React.FC<LanguageDropdownProps> = ({
       onChange={onChange}
       autoComplete={autoComplete}
       curvedWithShadow={curvedWithShadow}
+      icon={<SvgGlobe />}
```
```suggestion
      svgIcon={<SvgGlobe />}
```

tricaalexandru commented:
@philjonasfiltered there's a bit of a mismatch with the design. All in all is good, but here are the necessary changes

1. In the deisgne it's leaning to the left, in yours it's leaning to the right
2. There's a check mark for the current langauge
3. The hilight is of the entire cell, not changing the text of the value



<img width="269" alt="Screenshot 2020-12-08 at 13 00 26" src="https://user-images.githubusercontent.com/7725343/101475669-5a9a5800-3955-11eb-841a-091c1d0e8c3e.png">

tricaalexandru reviewed (changes requested):
@philjonasfiltered left comment in regards to the design

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -10,16 +11,25 @@ export interface DropdownProps {
 export interface DropdownPropsEnhanced extends DropdownProps {
   showDropdown: boolean
   toggleDropdown: () => void
+  isIconTightlyWrapped?: boolean
```
@philjonasfiltered I'm not sure what this prop is doing, perhaps explain and we can come up with a name that means what it's doing, it might be we could find a solution where we do not need this kind of thing.

philjonasfiltered replied:
@matthill8286 makes sense. Up until yesterday, this was a regular dropdown. Now with this new design, it's very far from what other dropdown look like. This is a zero-point task becoming a time sink 😞

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -64,22 +69,29 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
-          <StyledIcon width={16} height={16} rotate={showDropdown ? 270 : 90}>
-            <SvgEdit />
+          <StyledIcon
+            width={16}
+            height={16}
```
@philjonasfiltered If all icons in the designs are 16 x 16 then perhaps set that as a default value in the <Icon /> itself.

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -64,22 +69,29 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
+            width={16}
+            height={16}
+            isIconTightlyWrapped={isLanguageDropdown}
+            rotate={iconRotate !== undefined ? iconRotate : showDropdown ? 270 : 90}>
```
@philjonasfiltered Why not set iconRotate = false as a default, then you could do `iconRotate : showDropdown ? 270 : 90`

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -64,22 +69,29 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
+          <StyledList
+            active={!!showDropdown}
+            isSearchable={false}
+            isIconTightlyWrapped={isLanguageDropdown}>
```
```suggestion
            iconLeft={isLanguageDropdown}>
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -56,18 +61,20 @@ export const StyledTypo = styled(Typo)`
+    pointer-events: none;
+    position: absolute;
+    top: 50%;
+    ${isIconTightlyWrapped ? `left: 20%` : `right:  20px`};
```
```suggestion
    ${iconLeft ? `left: 20%` : `right:  20px`};
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -10,16 +11,25 @@ export interface DropdownProps {
 export interface ListProps {
   active: boolean
   isSearchable: boolean
+  isIconTightlyWrapped?: boolean
```
```suggestion
  iconLeft?: boolean
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -10,16 +11,25 @@ export interface DropdownProps {
+}
+
+export interface IconPropsEnhanced extends IconProps {
+  isIconTightlyWrapped?: boolean
```
```suggestion
  iconLeft?: boolean
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -30,6 +40,7 @@ export interface StyledDropdownProps {
 export interface StyledDropdownOptionProps {
   active: boolean
   onClick: React.MouseEventHandler<HTMLLIElement>
+  hasCheckmark?: boolean
```
```suggestion
  checkmark?: boolean
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -83,17 +90,22 @@ export const StyledList = styled.ol<ListProps>`
-  ${({ theme }) => `
-    display: block;
+  ${({ theme, hasCheckmark, active }) => `
+    display: ${hasCheckmark ? `inline-flex` : `block`};
```
```suggestion
    display: ${checkmark ? `inline-flex` : `block`};
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -83,17 +90,22 @@ export const StyledList = styled.ol<ListProps>`
 export const StyledDropdownOption = styled.li<StyledDropdownOptionProps>`
-  ${({ theme }) => `
-    display: block;
+  ${({ theme, hasCheckmark, active }) => `
```
```suggestion
  ${({ theme, checkmark, active }) => `
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -26,6 +26,10 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
+    svgIcon = <SvgEdit />,
+    iconRotate,
+    isSelectDisplayed = true,
+    isIconTightlyWrapped,
```
```suggestion
    iconLeft,
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -54,6 +58,7 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
   }
 
   const selectedItem = options.find((o, i) => i === currentIndex)
+  const isLanguageDropdown = iconRotate === 0
```
@philjonasfiltered What happens if another dropdown uses a icon thats at 0??

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -64,22 +69,29 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
       {listType === 'custom' ? (
         <StyledDropdown hasLabel={!!label} curvedWithShadow={curvedWithShadow}>
-          {currentIndex !== -1 && (
+          {isSelectDisplayed && currentIndex !== -1 && (
```
```suggestion
          {isOptionValueDisplayed && currentIndex !== -1 && (
```

matthill8286 commented on src/components/Molecules/Dropdown/Dropdown.tsx:
```diff
@@ -64,22 +69,29 @@ const DropdownWithoutState: FC<DropdownPropsEnhanced> = props => {
                 active={index === currentIndex}
                 label={item.label}
                 onClick={handleItemClick(index)}
+                hasCheckmark={isLanguageDropdown}
```
```suggestion
                checkmark={isLanguageDropdown}
```

philjonasfiltered commented on src/components/Molecules/Dropdown/Dropdown.interface.tsx:
```diff
@@ -10,16 +11,25 @@ export interface DropdownProps {
 export interface DropdownPropsEnhanced extends DropdownProps {
   showDropdown: boolean
   toggleDropdown: () => void
+  isIconTightlyWrapped?: boolean
+}
+
+export interface IconPropsEnhanced extends IconProps {
+  isIconTightlyWrapped?: boolean
```
This makes the container wrap the icon without any dead space to the left.

philjonasfiltered commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -56,18 +61,20 @@ export const StyledTypo = styled(Typo)`
+  ${({ active, isSearchable, theme, isIconTightlyWrapped }) => `
     min-width: 100%;
-    max-width: 100%;
+    ${isIconTightlyWrapped ? `` : `max-width: 100%;`}
```
this is breaking the tests

philjonasfiltered commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -83,17 +90,22 @@ export const StyledList = styled.ol<ListProps>`
     transform: ${active && 'scaleY(1)'};
     transform-origin: 0 0;
     transition: transform ${theme.transition.short} ease;
+    ${isIconTightlyWrapped ? `right: 0;` : ``}
```
this is breaking the tests

philjonasfiltered commented on src/components/Molecules/Dropdown/Dropdown.styled.ts:
```diff
@@ -83,17 +90,22 @@ export const StyledList = styled.ol<ListProps>`
     background: ${theme.color.white};
     hyphens: auto;
     text-align: left;
+    ${hasCheckmark && !active ? `padding-left: 45px;` : ``}
```
also breaking the tests. I need to figure out why...

tricaalexandru commented:
@philjonasfiltered it's way close but still needs a bit of love. 
https://projects.invisionapp.com/share/BUZB2DJGXCW#/screens/436779856

Missing:
1. The up/down arrow
2. Yours doesn't blend with the background (not sure if this matters. Maybe revisit when we have the header finalise)
3. Text is centered differently (@matthill8286 your call. Not sure how much it matters)
4. The scroll bar on right side.

matthill8286 commented:
> @philjonasfiltered it's way close but still needs a bit of love.
> https://projects.invisionapp.com/share/BUZB2DJGXCW#/screens/436779856
> 
> Missing:
> 
> 1. The up/down arrow
> 2. Yours doesn't blend with the background (not sure if this matters. Maybe revisit when we have the header finalise)
> 3. Text is centered differently (@matthill8286 your call. Not sure how much it matters)
> 4. The scroll bar on right side.

@tricaalexandru I'll check out the branch and run through it 👍

tricaalexandru commented:
@philjonasfiltered so Phil Cowans said we'll have a design review by Craig when the code is done. So we have to make it look identical to the design. I can still see some differences:

1, Either the arrow is too big or the globe is to small
2. The globe and arrow look Bolded
3. The box with the language has rounded corners
4. There is a space above the first item
5. The check icon should be smaller
6. The actual width should be bigger, but honestly I like your version better so let's go with that and see if Craig rejects it. 

@matthill8286 just FYI

philjonasfiltered commented:
> @philjonasfiltered so Phil Cowans said we'll have a design review by Craig when the code is done. So we have to make it look identical to the design. I can still see some differences:
> 
> 1, Either the arrow is too big or the globe is to small
Changed Globe size and icon
> 2. The globe and arrow look Bolded
Same as above
> 3. The box with the language has rounded corners
The design has square corners at the top and rounded at the bottom. I updated the component accordingly
> 4. There is a space above the first item
Removed the space
> 5. The check icon should be smaller
Check icon is smaller
> 6. The actual width should be bigger, but honestly I like your version better so let's go with that and see if Craig rejects it.
didn't change that 😄 
> @matthill8286 just FYI

tricaalexandru commented:
@philjonasfiltered what's the state of this?

philjonasfiltered commented:
> @philjonasfiltered what's the state of this?

If you and @matthill8286 are happy with it, I think it can be merged. If later we need adjustments I can always pick this work again.

matthill8286 commented:
@tricaalexandru @philjonasfiltered agreed we can always revisit it if it needs it, don't see the value of any more time on such a small thing
